### ISSUE TITLE
Segfault creating netlink sockets

### ISSUE THREAD

**Issue by nirvdrum** [oracle/truffleruby#3196]
I've finally traced a source of segfaults to the allocation of netlink sockets, which is used by the [raindrops gem](https://github.com/tmm1/raindrops/blob/1c18fd9c13f95fef6bcbdc0587d38886fa8e9064/ext/raindrops/linux_inet_diag.c#L111-L113). Given they're platform-specific, this is only a problem on Linux. Unfortunately, I've only been able to reliably reproduce the issue in a staging environment using Kubernetes. A simple reproduction is:

```
ruby -r socket -e 'loop { Socket.new(16, 524291, 4) }'
```

That should throw an `Errno::EMFILE` and it since it's uncaught, exit the application with an error. Here the exception is thrown but then the app segfaults. I don't have an hs_err log since writing out the log file apparently fails ("Can not save log file, dump to screen.."), but [here's what I see on the console](https://gist.github.com/nirvdrum/047ba9c0a81245a54a533b4738dc36ba). Running the code in a loop like this, the problematic frame consistently shows libjvmcicompiler.so. However, in my real deployment I usually see the error in a Ruby thread using NFI. It's possible they're two unrelated issues, but I think it's possible the attribution is based on what the system is doing at the time of the crash.

If I change the `loop` to `1_000_000.times` I see a `java.lang.OutOfMemoryError` instead.

This is with the TruffleRuby 23.0.0 release using the Oracle GraalVM on x86_64 Linux.

(I used a gist for the error report because the content is too long to put directly in the issue.)

EDIT: Tracked internally in GR-47764.

**Comment by eregon** [comment]
On CRuby 3.1.3:
```
$ ruby -v -rsocket -e 'loop { Socket.new(16, 524291, 4) }'             
ruby 3.1.3p185 (2022-11-24 revision 1a6b16756e) [x86_64-linux]
-e:1:in `initialize': Too many open files - socket(2) (Errno::EMFILE)
	from -e:1:in `new'
	from -e:1:in `block in <main>'
	from -e:1:in `loop'
	from -e:1:in `<main>'
```
If I run this on TruffleRuby Native, master or 23.0.0 it gives on my Linux machine:
```
$ ruby -v -rsocket -e 'loop { Socket.new(16, 524291, 4) }'      
truffleruby 23.0.0, like ruby 3.1.3, Oracle GraalVM Native [x86_64-linux]
<internal:core> core/errno.rb:48:in `handle': Too many open files - socket(2) (Errno::EMFILE)
	from /home/eregon/.rubies/truffleruby+graalvm-23.0.0/graalvm/languages/ruby/lib/truffle/socket/socket.rb:317:in `initialize'
	from -e:1:in `block in <main>'
	from <internal:core> core/kernel.rb:412:in `<unknown>'
	from <internal:core> core/kernel.rb:407:in `loop'
	from -e:1:in `<main>'
```

Using `ruby -v -rsocket -e 'i=0; loop { Socket.new(16, 524291, 4); p i+=1 }'` I see it goes until `524281`.
While CRuby 3.1.3 only goes until `1633`.

In JVM mode, I do see the segfault and it goes until `524252`:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f0c8a4302ad, pid=12377, tid=12377
#
# JRE version: Java(TM) SE Runtime Environment Oracle GraalVM 17.0.7+8.1 (17.0.7+8) (build 17.0.7+8-LTS-jvmci-23.0-b12)
# Java VM: Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 17.0.7+8.1 (17.0.7+8-LTS-jvmci-23.0-b12, mixed mode, sharing, tiered, jvmci, jvmci compiler, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C  [libjvmcicompiler.so+0x14302ad]
```

**Comment by eregon** [comment]
`Socket.new(16, 524291, 4)` is the same as `Socket.new(Socket::Socket::AF_NETLINK, Socket::SOCK_RAW|Socket::SOCK_CLOEXEC, 4)`.
The 4 is `NETLINK_INET_DIAG` but that's not exposed as a constant in Ruby.

It comes from the `raindrops` gem: https://github.com/tmm1/raindrops/blob/1c18fd9c13f95fef6bcbdc0587d38886fa8e9064/ext/raindrops/linux_inet_diag.c#L111-L113

**Comment by eregon** [comment]
Looking at `Socket#initialize` the only thing it really does is just calling `socket(2)` + `fcntl(fd, F_GETFL, 0)`:
https://github.com/oracle/truffleruby/blob/1c80de936d1c8de1e510157ad1d4e139985ff7fa/lib/truffle/socket/socket.rb#L310-L322

**Comment by eregon** [comment]
@nirvdrum Do you know if `raindrops` actually creates many sockets like that loop, i.e., does it actually run out of file descriptors? If so it sounds like that should be fixed somehow.

**Comment by nirvdrum** [comment]
It doesn't create sockets in a loop. I did it that way to cut out the _raindrops_ dependency and speed the process up. We have two ways to invoke _raindrops: as a background thread that periodically captures data and as a Rack middleware that sits around the request. Currently, I'm seeing this with the background thread collecting data once per second. After about 25 minutes, we crash the VM.

**Comment by nirvdrum** [comment]
Per @eregon's request, I ran with libgraal disabled (`--vm.XX:-UseJVMCINativeLibrary`) and there was no segfault. We see the `Errno::EMFILE` exception as expected and then the process subsequently exiting with status code 1.

**Comment by dougxc** [comment]
If the *raindrops* crash is different from the one in your reproducer, then having console output and a hs-err log will make investigating it significantly easier.

**Comment by christianhaeubl** [comment]
The crash happens during libgraal shutdown if the application ran out of file descriptors. We pushed fixes to both libgraal and Native Image, so this should be fixed in GraalVM 24.0.

**Comment by dougxc** [comment]
This will be fixed in a future release that incorporates these changes:
* https://github.com/oracle/graal/commit/d636f2f9e747de1ef547680b05d901e9633c4746
* https://github.com/openjdk/jdk/pull/15542
